### PR TITLE
Add lowering for BF16 arith/math operations to F32 with truncf/extf wrappers

### PR DESCRIPTION
**Pull request by rsuderman**
We can support BF16 execution by casting to F32 types and performing the F32 math in that space. This allows bf16 support once we have a simplified conversion betweeen bf16 and f32. For now this path is included after the BF16 to F32 promotion pass so it will not trigger. A follow up will fully enable BF16 support and disable the F32 conversion.

### CODE REVIEW

**Comment by rsuderman** [comment]
This CL cannot land until the upstream bf16 changes land.

**Review by jpienaar — COMMENTED** [review]
Nice!

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: Nit: Please add braces. (this diverges from LLVM style but is recommended in IREE)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:643]: I am wondering if this is enough. Arent the operations in the body `linalg` operations with `bf16` types? I am wondering why this is different compared to the `i1` -> `i8` handling?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (jpienaar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:643]: Do we/could we add a test that verifies this?

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp
Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: Fixed.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (rsuderman) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:643]: So unlikely the `i1` to `i8` conversion we do not change the underlying storage buffers, rather they remain as a 2-byte per element buffer. What does change is we perform casts surrounding all of the `arith` operations so that they hit known to work versions of the ops (e.g. `f32`). 

This occurs after the conversion to `scf` but should occur both within the dispatched for loops and the VM code, as we apply the conversion to any numerical operations throughout the model.

I am still waiting on the corresponding LLVM change to integrate as it missed the most recent integration cycle. Once it is in we should be able to introduce an execution example. For now we still perform the raw `bf32` to `i32` conversion on `mhlo` models.

Comment (rsuderman) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:643]: We may be able to do some similar tricks for `i1` packing as well. I imagine `exti` could have a similar conversion where we extract only the bits needed, though the loading semantics would need to be adjusted.

**Comment by rsuderman** [comment]
PR was updated with full support for CPU and enables BF16 natively for the CPU path.

VMVX requires some changes to include buffer / operation support for BF16 ops.

Vulkan and CUDA likely require additional work, migration can be performed once we settle this is a good idea.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:643]: ok I see... well it makes me a bit concerned that we have two places where we are adjusting types... but it is true, one is at buffer level and another is at register level....

**Review by MaheshRavishankar — APPROVED** [review]
Looks good! 

@powderluv I think this will enable running stable diffusion on CPU backends now.

File: tests/e2e/xla_ops/BUILD.bazel
Comment (jpienaar) [inline tests/e2e/xla_ops/BUILD.bazel]: Should we just remove here?

**Review by jpienaar — APPROVED** [review]

File: tests/e2e/xla_ops/BUILD.bazel
Comment (rsuderman) [reply tests/e2e/xla_ops/BUILD.bazel]: Removed.

**Pull request merged**
